### ISSUE TITLE
PhantomJS Hangs on Certain Sites with Custom Scripts

### ISSUE THREAD

**Issue by dylotribal** [jonnnnyw/php-phantomjs#74]
The below works for some sites like google.com, but not others like this one:
http://www.buzzfeed.com/johnc4197e36ba/lonely-bridget-still-lonely-16imt

... hangs almost like phantom.exit() is not called...

//
// scroll_to_bottom.proc
//
var page  = require('webpage').create();
page.settings.resourceTimeout = 10000;

var response; 
page.onResourceReceived = function (r) {
	response = r;
};

page.onResourceError = function(resourceError) {
	page.reason = resourceError.errorString;
	page.reason_url = resourceError.url;
	console.log(JSON.stringify({ content: "error:"+page.reason + ", reason url: " + page.reason_url }));
	phantom.exit();
};

page.open ('{{ request.getUrl() }}', '{{ request.getMethod() }}', '{{ request.getBody() }}', function (status) {
	
	console.log(JSON.stringify({ content: page.content }));
	phantom.exit();
	
});

page.onResourceTimeout = function(request) {
 phantom.exit();
};

phantom.onError = function(msg, trace) {
	phantom.exit(1);
};

//
// PHP used:
//
use JonnyW\PhantomJs\Client;
use JonnyW\PhantomJs\DependencyInjection\ServiceContainer;

$fileName = 'scroll_to_bottom';
$filePath = '/var/www/sites/scripts/';
$serviceContainer = ServiceContainer::getInstance();
$procedureLoader = $serviceContainer->get('procedure_loader_factory')->createProcedureLoader($filePath);

$client = Client::getInstance();
$client->debug(true);
$client->addOption('--ssl-protocol=tlsv1');
$client->addOption('--load-images=false');
//$client->addOption('--ignore-ssl-errors=true');
//$client->addOption('--web-security=false');
$client->getProcedureLoader()->addLoader($procedureLoader);
$client->setPhantomJs('/usr/bin/phantomjs');
$client->setBinDir('/var/www/sites/scripts/bin');

$JSRequest  = $client->getMessageFactory()->createRequest();
$JSRequest->setTimeout(10);

$JSRequest->setType($fileName); 
$JSRequest->setMethod('GET');
$JSRequest->setUrl("http://www.buzzfeed.com/johnc4197e36ba/lonely-bridget-still-lonely-16imt");
//$JSRequest->setUrl("http://www.google.ca");

$JSResponse = $client->getMessageFactory()->createResponse();
$client->send($JSRequest, $JSResponse);
//echo $client->getLog();
//var_dump($JSResponse->getConsole());

echo $JSResponse->content;

**Comment by jonnnnyw** [comment]
My guess would be something to do with the phantomloader script which was a work around anyway. Version 4 has just been released and that phantomloader file has been deprecated so see if upgrading solves your issues.

**Comment by JalfResi** [comment]
Have a similar issue with the URL https://www.intrepidtravel.com/au/morocco/morocco-family-holiday-93373.

In our investigations we discovered that it appears that Procedure/Procedure.php:112 is where the script stalls - it appears that stream_get_contents is expecting more data and hangs there, even though the PhantomJS process has |completed" output content.

We were able to work around this by modifying the proc_open cmd executed and redirecting the output of the phantomjs process to temporary files, which we read back in place of stream_get_contents.

I am experimenting with resolving this properly (going to try some alternatives such as selecting over the streams and reading in chunks, with a timeout) and will update this issue with my progress.
